ls1intum/Artemis
# Allow Locking and Unlocking Programming Exercises

### Checklist

### Motivation and Context
Allow instructors to lock and unlock all programming exercise repositories of a particular exercise.

### Description
- Add two REST endpoints (course instructors only)
- Add two buttons in the client in the programming exercise details view (course instructors only)
- Provide modals with warnings and explanations
- Add tests for woth endpoints
- Fix potential security issues with notifications, we only allow ADMIN for system notifications now
- Fix, add and improve notification tests
- Fix small documentation issue in the student exam TS docs

### Steps for Testing
0. Programming exercise in course with multiple participations needed
1. Try both **locking** and **unlocking** the participations and check that you:
2. Get correct messages, warning modals and translations in DE and EN
3. Then get a message that the server is working on the operation
4. You get a notification inside Artemis as an instructor, that the operation was successful 
    (It looks you might need to reload the page due to the notification system working suboptimally, see #2691)
5. View the student repositories in Bitbucket and check that under "Settings" (gear at the bottom) => Repository Permissions, the write permission is enabled/disabled depending on the operation.

### Test Coverage
 - ProgrammingExerciseResource 82%
 - NotificationResource 100%
 - SystemNotificationResource 100%
 - ProgrammingExerciseService 82%

### Screenshots
![image](https://user-images.githubusercontent.com/11130248/104644683-5f69ce80-56ae-11eb-88b9-54c153cdc92a.png)

![image](https://user-images.githubusercontent.com/11130248/104645847-ed928480-56af-11eb-92da-4edd8c78b30d.png)

![image](https://user-images.githubusercontent.com/11130248/104645971-1ca8f600-56b0-11eb-81cd-53adbf15c46c.png)
